### DISCUSSION TITLE
mock 을 사용한 단위 테스트 vs 통합 테스트

### DISCUSSION THREAD

**Discussion by nayonsoso** [solid-connection/solid-connect-server#136] [논의]
서비스 코드를 테스트할 때 어떤 테스트가 적합할지 이야기해봅시다.

**Comment by Gyuhyeok99** [comment]
mock을 사용한 단위테스트가 필요할 때도 있고 통합테스트가 필요할 때도 있다고 생각합니다. 그래서 테스트 피라미드 원칙에 따라, 단위 테스트와 통합 테스트를 조합해 사용하되, 상황에 따라 적합한 테스트를 선택하는 방식은 어떨까요?

<img width="719" alt="test피라미드" src="https://github.com/user-attachments/assets/f139d44e-75b6-457d-be4e-344114c1baf5" />


서비스 계층은 비즈니스 로직을 처리하는 역할을 하고 컨트롤러에서 받은 요청을 처리하고, 필요한 데이터를 데이터베이스 또는 다른 외부 서비스에서 가져오고, 결과를 반환합니다. 따라서 서비스 계층은 로직 검증과 연동 검증 두 가지가 모두 중요하다고 생각합니다.

단위 테스트:
특정 비즈니스 로직이나 알고리즘의 정확성을 검증해야 할 때.
외부 의존성이 없는 클래스나 메서드의 동작을 독립적으로 테스트할 때.
Mock을 통해 빠르게 검증할 수 있는 경우.

통합 테스트:
여러 계층(Service, Repository, Controller)이 연결된 흐름을 테스트해야 할 때.
데이터베이스나 외부 API와의 실제 연결을 테스트해야 할 때.
종단 간(E2E) 테스트가 필요하지만, 그보다 가벼운 검증을 목표로 할 때.

아직 구체적으로 어떻게 나눠야할지 생각은 못했는데 다른 분들 의견도 듣고싶습니다!

**Comment by nayonsoso** [comment]
저는 서비스 코드를 테스트할 때 통합테스트를 선호합니다😊
예를 들어 설명해보겠습니다.

---

```java
@Test
void 닉네임을_수정한다() {
    // given
    NicknameUpdateRequest request = new NicknameUpdateRequest("newNickname");
    given(siteUserRepository.getByEmail(siteUser.getEmail())).willReturn(siteUser);

    // when
    NicknameUpdateResponse response = siteUserService.updateNickname(siteUser.getEmail(), request);

    // then
    assertThat(response.nickname()).isEqualTo(siteUser.getNickname());
}
```

위 테스트 코드는 `siteUserService.updateNickname()`를 검증하는 코드입니다.
저는 여기에서 `given(siteUserRepository.getByEmail(siteUser.getEmail())).willReturn(siteUser);` 부분이 불필요하다고 생각합니다.

테스트 코드에서는 검증하고자 하는 것이 명확해야 합니다.
목적과 관련없는 코드는 무엇을 테스트하고자 하는지를 흐리게 만들 수 있습니다.
위 테스트 코드는의 목적은, **[닉네임을 수정했을 때, 이것이 잘 수정되었는지]** 를 테스트하는 것입니다.
이를 생각해봤을 때, 테스트코드를 읽는 사람 입장에서는 
siteUserRepository.getByEmail 이 호출된다는 것과, 이를 호출했을 때 siteUser 가 반환되어야 한다는 사실을 알 필요가 없습니다.

---

그래서 이렇게 바뀌어야 한다고 생각하는데요👇

```java
@Test
void 닉네임을_수정한다() {
    // given
    siteUserRepository.save(siteUser);
    NicknameUpdateRequest request = new NicknameUpdateRequest("newNickname");

    // when
    NicknameUpdateResponse response = siteUserService.updateNickname(siteUser.getEmail(), request);

    // then
    assertThat(response.nickname()).isEqualTo(siteUser.getNickname());
}
```

given().willReturn() 대신에 `siteUserRepository.save(siteUser);` 부분이 추가되었습니다.
이는 '닉네임을 바꾸기 전에, 회원이 등록되어있어야 한다'는 전제 조건을 더 직관적으로 알려준다고 생각합니다.
그리고 이를 위해서는 단위테스트가 아니라 통합 테스트가 되어야 합니다!

**Reply by Gyuhyeok99** [reply]
통합 테스트가 확실히 더 직관적인 거 같습니다. 실행 시간이 조금 늘어나긴 하지만, 코드가 명확해져서 코드 리뷰 시간도 줄어들고 협업 시 장점이 더 많은 것 같습니다! 테스트컨테이너 도입 장점이 또 느껴지네요!

**Comment by nayonsoso** [comment]
추가로, 저는 단위 테스트를 위해서 아주 긴 stub 코드가 작성되는걸 선호하지 않습니다😞
규혁님도 저희 테스트 코드 중에서 given 절이 아주 긴~ 코드를 보신적이 있으실거예요..
이런 코드는 무엇을 위한 테스트코드인지 파악하기도 어렵고, 테스트 코드를 읽는데에 피로감마저 들게 하더라고요.

그리고 서비스 코드를 단위 / 통합으로 나누는 것도 리소스라는 생각이 들어서..
 서비스 코드들을 테스트할 때는 통합테스트로 통일하는게 어떤지 제안드려봅니다!

**Reply by Gyuhyeok99** [reply]
넵 좋습니다! 그럼 이전에 작성한 단위테스트 PR은 닫을까요?

**Reply by nayonsoso** [reply]
닫고 새로 PR 올리시는게 좋을 것 같습니다 😭

**Comment by nayonsoso** [comment]
# 👨🏻‍⚖️ 결정된 내용

### 서비스 함수에 대한 테스트 코드는 **레포지토리를 포함한 통합 테스트**로 작성한다.

서비스 함수를 단위 테스트하기 위해서는 내부에서 호출하는 모든 함수들을 stub 해줘야 한다.
길어지는 stub 부분은 테스트의 목적이 무엇인지 불명확하게 한다.
또한 서비스 함수의 세부 구현에 의존하는, 변경에 유연하지 못한 테스트 코드가 된다.
이를 지양하기 위해 통합 테스트로 작성한다.
